### ISSUE TITLE
Replace KeyValuePair with Tuple

### ISSUE THREAD

**Issue by isaacabraham** [Microsoft/Mobius#340]
Just noticed in the examples that KVP is used for many of the "pairs" - not just in the samples but within method signatures e.g. GroupBy. Perhaps these could (should?) be replaced with Tuples.
- Native support in F# already e.g.

``` csharp
var wordCounts = words.Map(w => new KeyValuePair<string, int>(w.Trim(), 1))  
                      .ReduceByKey((x, y) => x + y);  
```

becomes

``` fsharp
let wordCounts = words.Map(fun w -> w.Trim(), 1)
                      .ReduceByKey(fun (x, y) -> x + y)
```
- Next version of C# will probably have similar syntax. Even with C#6 the experience is nicer: -

``` csharp
var wordCounts = words.Map(w => Tuple.Create(w.Trim(), 1))  
                      .ReduceByKey((x, y) => x + y);  
```
- More importantly, Tuple is a well known type across many languages (including Scala) and I think better represents what the intent is.

**Comment by vasily-kirichenko** [comment]
:+1:

**Comment by isaacabraham** [comment]
Could someone just give high-level opinion on this change? I'd be happy to make this change myself, it's hopefully just a boilerplate menial change - but there's no point my spending time on it if there's no agreement that it has value.

**Comment by vasily-kirichenko** [comment]
If that `KeyValuePair'2` is a struct, then we should not replace it with `Tuple'2`, because it's a referential type.

**Comment by isaacabraham** [comment]
Why?

**Comment by vasily-kirichenko** [comment]
performance

**Comment by isaacabraham** [comment]
premature optimization? :-)

**Comment by vasily-kirichenko** [comment]
As it seems to be created on each data piece (line, record, whatever), there will be very large number of unnecessary allocation.

**Comment by isaacabraham** [comment]
I'm thinking the cost of GC allocation would probably be massively outweighed by the IO costs normally associated with a distributed data application like Spark. At least in work I've done before, that's been the case.
________________________________

**Comment by skaarthik** [comment]
Replacing KVP with Tuple will achieve conceptual parity with scala and make the code a bit less verbose as @isaacabraham has pointed out but the performance/memory implications may not be worth it as called out by @vasily-kirichenko. With the worker implementation getting multithreaded (https://github.com/Microsoft/Mobius/pull/349) we need to be careful with every bit of overhead especially in streaming workloads.

The other consideration is to think about changes to Mobius public API and the implementation in https://github.com/Microsoft/Mobius/blob/master/csharp/Adapter/Microsoft.Spark.CSharp/Core/PairRDDFunctions.cs. 

If we can validate that the performance and memory usage in KVP and Tuple implementations are comparable, we could consider making this change along with other major changes that will happen with Spark 2.0 upgrade.

**Comment by isaacabraham** [comment]
@skaarthik OK - what sort of performance and memory metrics would you like to see to validate this?

**Comment by skaarthik** [comment]
Thanks @isaacabraham. 
Mobius perf suite can be used for validating if the memory usage and run time are comparable. Check out RunRDDMaxDeletionsByUser perf test @ https://github.com/Microsoft/Mobius/blob/master/csharp/Perf/Microsoft.Spark.CSharp/FreebaseDeletionsBenchmark.cs#L47. It uses KVP and ReduceByKey.

Once the PR is ready I can also perform the validation using a streaming workload we have at Microsoft to test UpdateStateByKey implementation that uses KVP. https://github.com/Microsoft/Mobius/blob/master/csharp/Adapter/Microsoft.Spark.CSharp/Streaming/PairDStreamFunctions.cs#L342

**Comment by isaacabraham** [comment]
Is there a reason for so little use of var in the codebase? Even though C# has somewhat limited type inference, even using that a little more would help with both readability and refactoring. I've been looking into this briefly this evening and it's been an absolute pain to replace the types on both sides of the equals.

**Comment by vasily-kirichenko** [comment]
I'm sure it can be done in one click with ReSharper.

**Comment by isaacabraham** [comment]
@vasily-kirichenko or maybe a few "one click" :-)

**Comment by vasily-kirichenko** [comment]
@isaacabraham R# can apply any refactoring to the whole file, project and even solution.

**Comment by isaacabraham** [comment]
FYI... http://vaskir.blogspot.de/2016/05/upcoming-f-struct-tuples-are-they.html?m=1

**Comment by abatishchev** [comment]
Any update on this, please?

**Comment by dwnichols** [comment]
I have a active pull request open which addresses this: #605.

**Comment by abatishchev** [comment]
Oh, yes, indeed. Posted this comment before seeing those two prs. Does it make sense to close the issue to do not fork the discussion on this matter?

**Comment by skaarthik** [comment]
Merged b95520eff8ea63914fd58bb83a932d1015bcce46
